Feature: Chat message feedback

0xi4o opened the pull request:
- Adds a dialog where chat feedback settings can be controlled.
- Adds feedback option to chat messages in embed - thumbs up or thumbs down. When one of these are clicked, an input dialog is show where users can leave written feedback.
- Shows feedback in `View Messages` dialog. 

![flowise-chat-feedback-settings](https://github.com/FlowiseAI/Flowise/assets/4386534/f4cf2770-f518-4575-b36b-6951c5e3ea7c)
![flowise-chat-feedback-settings-2](https://github.com/FlowiseAI/Flowise/assets/4386534/1030d2b3-c1f2-4d89-9701-5197028c90c7)
![flowise-chat-feedback-view-messages](https://github.com/FlowiseAI/Flowise/assets/4386534/b369f9a5-2157-4bd4-97f7-231048ef2419)

HenryHengZJ commented on packages/server/src/index.ts:
```diff
@@ -1460,6 +1534,28 @@ export class App {
+                .andWhere(sessionId ? 'chat_message.sessionId = :sessionId' : 'TRUE', {
+                    sessionId: sessionId ?? (chatId ? IsNull() : undefined)
+                })
+                .andWhere(fromDate && toDate ? 'chat_message.createdDate = :createdDate' : 'TRUE', {
```
filtering by date is not working anymore, please check.
<img width="881" alt="image" src="https://github.com/FlowiseAI/Flowise/assets/26460777/526eda58-bab4-4f7e-9a26-55e109162078">

0xi4o replied:
Fixed this issue.

HenryHengZJ commented on packages/server/src/index.ts:
```diff
@@ -1460,6 +1534,28 @@ export class App {
+                .createQueryBuilder('chat_message')
+                .leftJoinAndMapOne('chat_message.feedback', ChatMessageFeedback, 'feedback', 'feedback.messageId = chat_message.id')
+                .where('chat_message.chatflowid = :chatflowid', { chatflowid })
+                .andWhere(chatType ? 'chat_message.chatType = :chatType' : 'TRUE', { chatType })
```
question: what is the `TRUE` here?

0xi4o replied:
If a value doesn't exist, `TRUE` will return the value as if there wasn't a filter. Anyway, I've updated how the query is built for optional filters so it's clearer.

HenryHengZJ commented on packages/server/src/index.ts:
```diff
@@ -599,10 +605,76 @@ export class App {
             return res.json(results)
         })
 
+        // ----------------------------------------
```
@ocflowiseai also new endpoints that needs to be updated once its merged - https://github.com/FlowiseAI/Flowise/pull/1806

HenryHengZJ commented on packages/ui/src/ui-component/dialog/ViewMessagesDialog.js:
```diff
@@ -449,7 +463,33 @@ const ViewMessagesDialog = ({ show, dialogProps, onCancel }) => {
+                            marginRight: 8
+                        }}
+                    >
+                        <StatsCard title='Total Messages (External)' stat={`${stats.totalMessages}`} />
```
what does the (External) means?

0xi4o replied:
Embed since feedback is not available in the UI. I'll change the text.

saatchi-david commented:
This is really cool. Curious if there's an ETA, and if the data can be passed to Langsmith?

0xi4o commented:
@saatchi-david Should be available in the next release. We only have our own implementation at the moment. Passing the feedback to langsmith will come in a future release.

0xi4o commented:
@HenryHengZJ I've added the feedback functionality to the UI too. Also updated the `ViewMessagesDialog` - date and source filters will apply to stats too now.

![Flowise-Low-code-LLM-apps-builder (7)](https://github.com/FlowiseAI/Flowise/assets/4386534/910178e5-0081-4a43-ad89-496837d71e25)
![Flowise-Low-code-LLM-apps-builder (6)](https://github.com/FlowiseAI/Flowise/assets/4386534/3224eb67-3adf-4719-b102-508609f673d6)

saatchi-david commented:
Awesome, thanks @0xi4o!

HenryHengZJ commented on packages/ui/src/ui-component/dialog/ChatFeedbackContentDialog.jsx:
```diff
@@ -0,0 +1,76 @@
+        else dispatch({ type: HIDE_CANVAS_DIALOG })
+        return () => {
+            dispatch({ type: HIDE_CANVAS_DIALOG })
+            setFeedbackContent('')
```
always remember to clear the state when we exit the dialog
